File: Data Structures & Algorithms/depth-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:    return 0
        stk = deque([[root, 1]])
        height = 0

        while stk:
            node, depth = stk.pop()
            height = max(height, depth)
            if node.left:
                stk.append([node.left, depth + 1])
            if node.right:
                stk.append([node.right, depth + 1])
        
        return height

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    void dfs(TreeNode *root, pair<int, int> &pr) {
        if (!root)
            return;
        
        dfs(root->left, pr);
        // if (pr.first == 0)
        //     return;
        
        pr.first --;
        if (pr.first == 0) {
            pr.second = root->val;
            return;
        }

        dfs(root->right, pr);
    }
public:
    int kthSmallest(TreeNode* root, int k) {
        pair<int, int> pr = {k, -1};
        dfs(root, pr);
        
        return pr.second;
    }
};
